Make genes and GPR consistent with SBML spec and add context manager to GRPs

akaviaLab opened the pull request:
* [X] fix #1183 #1185 #1157
* [X] description of feature
This removes the function gene.remove_from_model() since that function was old and weird
* [X] description of fix
Fixes the problems with copying, fixes inconsistent and inappropriate removal of genes (#1185) and started adding context to reaction._update_genes_from_gpr()

I haven't been able to add context to reaction.gene_reaction_rule (setter). If someone can add to that please, I'd appreciate it.

## New behavior

As a reference here the behavior with those changes.

<details>

```
In [1]: from cobra.test import create_test_model

In [2]: mod = create_test_model("textbook")

In [3]: mod.reactions[0].genes
Out[3]: frozenset({<Gene b0351 at 0x7f6d17a1b4f0>, <Gene b1241 at 0x7f6d17a1b460>})

In [4]: mod.reactions[0].gpr
Out[4]: cobra.core.gene.GPR('b0351 or b1241')

In [5]: mod.reactions[0].gene_reaction_rule = "b0351"

In [6]: mod.reactions[0].genes
Out[6]: frozenset({<Gene b0351 at 0x7f6d17a1b4f0>})

In [7]: mod.reactions[0].gpr
Out[7]: cobra.core.gene.GPR('b0351')

In [8]: mod.genes.b1241
Out[8]: <Gene b1241 at 0x7f6d17a1b460>

In [9]: mod.genes.b1241.reactions
Out[9]: frozenset({<Reaction ALCD2x at 0x7f6d179d2f40>})

In [10]: mod.reactions[0]
Out[10]: <Reaction ACALD at 0x7f6d17a1b490>

In [11]:                                                                                                                                                                      
Do you really want to exit ([y]/n)? y
 ~/code/cobrapy   fix_1185  ipython
Python 3.9.10 (main, Feb 22 2022, 13:54:07) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cobra.test import create_test_model

In [2]: mod = create_test_model("textbook")

In [3]: mod.reactions[0]
Out[3]: <Reaction ACALD at 0x7f5921403550>

In [4]: mod.genes.b1241.reactions
Out[4]: 
frozenset({<Reaction ACALD at 0x7f5921403550>,
           <Reaction ALCD2x at 0x7f59213bc880>})

In [5]: mod.reactions[0].gpr
Out[5]: cobra.core.gene.GPR('b0351 or b1241')

In [6]: mod.reactions[0].gene_reaction_rule = "b0351"

In [7]: mod.reactions[0].genes
Out[7]: frozenset({<Gene b0351 at 0x7f59214035b0>})

In [8]: mod.reactions[0].gpr
Out[8]: cobra.core.gene.GPR('b0351')

In [9]: mod.genes.b1241
Out[9]: <Gene b1241 at 0x7f5921403520>

In [10]: mod.genes.b1241.reactions
Out[10]: frozenset({<Reaction ALCD2x at 0x7f59213bc880>})

In [11]: len(mod.genes)
Out[11]: 137

In [12]: for r in mod.reactions:
    ...:     r.gene_reaction_rule = ""
    ...: 

In [13]: len(mod.genes)
Out[13]: 137

In [14]: from cobra.io import read_sbml_model, write_sbml_model

In [15]: write_sbml_model(mod, "textbook_no_gprs.xml")

In [16]: read = read_sbml_model("textbook_no_gprs.xml")

In [17]: mod.reactions[0].genes
Out[17]: frozenset()

In [18]: read.reactions[0].genes
Out[18]: frozenset()

In [19]: len(mod.genes)
Out[19]: 137

In [20]: len(read.genes)
Out[20]: 137

```

</details>

Commit:
Removed the function Gene.remove_from_model() in gene.py

Commit:
reaction.py reaction._update_genes_from_gpr() will not remove genes from model

Commit:
fix error in _update_genes_from_gpr()

Commit:
try to minimize usage of reaction._genes. Should be using _update_genes_from_gpr() when possible.

Commit:
actually not removing genes from model

Commit:
modify test appropriately to not removing genes

Commit:
added context, made sure that when removing gene from model gene._model is set to None

Commit:
isort and black

Commit:
isort and black again

akaviaLab commented:
Grumble.
If I turn this
from typing import (
     TYPE_CHECKING,
      FrozenSet,
      Iterable,
      Optional,
      Set,
     Tuple,
     Union,
)

To
from typing import TYPE_CHECKING, FrozenSet, Iterable, Optional, Set, Tuple, Union

Like isort requires, black on my computer complains. I can try it and see if black on the CI will complain, but I think it will.

cdiener commented on src/cobra/core/reaction.py:
```diff
@@ -590,46 +589,65 @@ def genes(self) -> FrozenSet:
+                        # Maybe should be
+                        # context(partial(self._model.genes.__isub__, [new_gene]))
+                new_gene = model_genes.get_by_id(g_id)
+                self._genes.add(new_gene)
```
use `self._associate_gene(new_gene)` which will perform the steps in lines 633-634.

akaviaLab replied:
I left it as is, because the below comment will do it, and I do need the new_genes set for comparison. In this case, it will try to add new genes to rxn._genes twice, but that doesn't matter since it is a set.

akaviaLab replied:
Sorry, I could have deleted self._genes.add(new_gene) and left the other. If you'd like to delete that, please delete the context below it. Or I could do it.

akaviaLab replied:
Not sure how it can be fixed, since I need self._genes in line 637, and I need new_genes. Maybe it can be revised somehow, but I'm not sure. The ways I tired just led to errors.

cdiener commented on src/cobra/core/reaction.py:
```diff
@@ -590,46 +589,65 @@ def genes(self) -> FrozenSet:
 
         # Make the genes aware that it is involved in this reaction
         for g in self._genes:
             g._reaction.add(self)
```
use `self._associate_gene(new_gene)`

akaviaLab replied:
Done

cdiener commented on src/cobra/core/reaction.py:
```diff
@@ -590,46 +589,65 @@ def genes(self) -> FrozenSet:
-                    warn(f"could not remove old gene {g.id} from reaction {self.id}")
+        for g in old_genes.difference(new_genes):
+            try:
+                g._reaction.remove(self)
```
use `self._disassociate_gene(g)`.

akaviaLab replied:
Done

Commit:
ran isort, no black

Commit:
trying black again

Commit:
trying black again, with updated black

synchon commented:
@akaviaLab Hi! I believe you are facing the above problem multiple times on your local system. My suggestion:
1. Update all packages in your development Python virtual environment and importantly the development ones.
2. Then, try to run `isort` and `black`.

I believe this will solve it for you. If not, there might be something else involved.

cdiener commented:
I think the timeouts are too strict @Midnighter. Seems to lead to a lot of random CI failures.

Commit:
maybe gene_reaction_rule has context now

cdiener reviewed (changes requested):
Awesome looks great. Just some minor changes to use the build in functions.

Commit:
corrected as suggested

Commit:
Revert "maybe gene_reaction_rule has context now"

This reverts commit ce3f1e1670523651e78c0aa5426a26f14eab98fc.

akaviaLab reviewed (commented):
I made the changes, mostly as suggested.
I tried adding context to update_genes_from_gpr(), but it always failed when trying to test it. I'm doing something wrong, I think. If you can improve on the context, that would be great.

cdiener committed:
enable context manager for GPRs

cdiener commented:
Try this one. We probably still need tests for the context manager, but looks okay:

```
In [1]: from cobra.test import create_test_model

In [2]: from cobra.core.gene import GPR

In [3]: mod = create_test_model("textbook")

In [4]: with mod:
   ...:     mod.reactions[0].gpr = GPR.from_string("blub")
   ...:     print(mod._contexts[0]._history)
   ...: 
[functools.partial(<function Reaction.gpr at 0x7f6096a45310>, <Reaction ACALD at 0x7f6092da6850>, cobra.core.gene.GPR('b0351 or b1241')), functools.partial(<function remove_genes at 0x7f6096a349d0>, model=<Model e_coli_core at 0x7f60cd83fc70>, gene_list=[<Gene blub at 0x7f60cd7cde20>], remove_reactions=False), functools.partial(<built-in function setattr>, <Gene blub at 0x7f60cd7cde20>, '_model', None), functools.partial(<bound method Reaction._dissociate_gene of <Reaction ACALD at 0x7f6092da6850>>, <Gene blub at 0x7f60cd7cde20>), functools.partial(<bound method Reaction._associate_gene of <Reaction ACALD at 0x7f6092da6850>>, <Gene b1241 at 0x7f6092da6820>), functools.partial(<bound method Reaction._associate_gene of <Reaction ACALD at 0x7f6092da6850>>, <Gene b0351 at 0x7f6092da68b0>)]

In [5]: mod.reactions[0].gpr
Out[5]: cobra.core.gene.GPR('b0351 or b1241')

In [6]: mod.reactions[0].genes
Out[6]: frozenset({<Gene b0351 at 0x7f6092da68b0>, <Gene b1241 at 0x7f6092da6820>})

In [7]: mod.genes.blub
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File ~/code/cobrapy/src/cobra/core/dictlist.py:526, in DictList.__getattr__(self, attr)
    525 try:
--> 526     return DictList.get_by_id(self, attr)
    527 except KeyError:

File ~/code/cobrapy/src/cobra/core/dictlist.py:74, in DictList.get_by_id(self, id)
     73 """Return the element with a matching id."""
---> 74 return list.__getitem__(self, self._dict[id])

KeyError: 'blub'

```

cdiener committed:
remove TODO and warning

akaviaLab commented:
> Try this one. We probably still need tests for the context manager, but looks okay:
> 
Skipped for brevity. 
Thank you very much!
Would you like me to try and add the various tests for context and GPR, or should that be a different PR?

cdiener commented:
That would be great. Definitely can add them here.

Commit:
added some context to test_core_reaction.py and model.py

Commit:
fixed some tests, renamed gpr to rules in reaction.__iadd for tidiness

Commit:
added context to gpr modifications in delete.py:remove_genes and appropriate test

Commit:
added context to gpr modifications in modify.py:rename_genes and appropriate test

cdiener commented on src/cobra/manipulation/delete.py:
```diff
@@ -344,14 +346,24 @@ def remove_genes(
             remover.visit(rxn.gpr)
+            # If the remover completely removed the AST tree from the GPR, it will not
+            # have body at all, which is why this isn't if body is None.
             if "body" not in rxn.gpr.__dict__.keys():
```
In that case `in rxn.gpr.__dict__` would be enough but I think `hasattr(rxn.gpr, "body")` is more pythonic.

akaviaLab replied:
Done

cdiener commented on src/cobra/manipulation/modify.py:
```diff
@@ -169,6 +173,10 @@ def rename_genes(model: "Model", rename_dict: Dict[str, str]) -> None:
             model.genes._dict.pop(gene.id)  # ugh
```
Very informative comment here ;) I think we can remove the comment.

akaviaLab replied:
Done. Is there a better way to do this, without pop and index and so on?

cdiener replied:
`model.genes.remove(gene.id)` should work. That also fixes the (numeric) indices. `model.reactions/genes/metabolites` are all [DictLists](https://github.com/opencobra/cobrapy/blob/devel/src/cobra/core/dictlist.py).

akaviaLab replied:
Done

Commit:
minor error correction in delete.py
renamed _update_genes_from_gpr to update_genes_from_gpr because it used outside of Reaction.

Commit:
black again

cdiener reviewed (commented):
Some small comments.

Commit:
corrected by comments

akaviaLab reviewed (commented):
Done. What do you think about the context and context tests added?

Commit:
replaced pop with remove

cdiener commented on src/cobra/manipulation/modify.py:
```diff
@@ -166,9 +170,13 @@ def rename_genes(model: "Model", rename_dict: Dict[str, str]) -> None:
             gene.id = new_name
             model.genes[gene_index] = gene
+            if context:
+                model.genes.remove(gene.id)
```
Shouldn't those be inside a `context(...)` call. Looks odd that having a context should trigger an immediate removal.

akaviaLab replied:
You are correct. I tried setting up the context to be context(partial()) call, but I'm not having any luck. If you can review this section, and fix it, I'd appreciate it.
Not sure how to do this correctly.

akaviaLab replied:
Using remove leads to it failing the test earlier.

cdiener replied:
Okay, will take a look, though we don't have to ensure that the (numeric) index is reset correctly when reverting a gene deletions. Reaction deletions don't do that either.

akaviaLab replied:
We might have to reset the numeric index in some way, because asking 
```
if 'foo' in model.genes
```
relies on the model.genes._dict, so partially on the index, I think.

cdiener replied:
Yeah I meant only the numeric index. So model.genes[id] == gene has to be true but if it was the 5th entry before it can now be the 21st entry (position does not matter).

cdiener reviewed (commented):
One small doubt about an `if context:` block...

Commit:
trying to fix context in rename_genes

cdiener committed:
fix rename_genes context

cdiener commented:
Context manager is always executed from last to first (like a rewind), so that can sometimes be tricky.

akaviaLab commented:
I think this is pretty much done. Can we merge?
